Permit "throw new SomeException();" in @Pure methods; fixes #951

mernst opened the pull request:
Code-review this after merging #1670

smillst commented on dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java:
```diff
@@ -183,15 +183,49 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void ignore) {
-                purityResult.addNotDetReason(node, "object.creation");
-            } else {
-                purityResult.addNotBothReason(node, "object.creation");
+            // The expression in "throw EXPR;" is allowed to be non-@Deterministic, so long as it is
```
I was very confused when reading this comment because it is about throw expressions, but this method is `visitNewClass`.   It would be clearer to move this comment and the implementation to `visitThrow`.  (Just check if the thrown expression is a NewClassTree and if it is side effect free.)

mernst replied:
Why is it necessary to check if the NewClassTree is side-effect-free?

mernst replied:
The visitor must visit the arguments to the `new` expression, because one of its arguments might have a side effect.  Therefore, even if this code were moved into `visitThrow`, then there would still need to be a special case in `visitNew`.  (Or the code in throw would have to iterate over and visit each of the arguments to `new`; is that what you are suggesting?)

smillst replied:
Oh right, the NewClassTree doesn't have to be side effect free, but the arguments should be.

smillst replied:
If you move the code to visitThrow, then it would need to iterate over the arguments.  Which would make it possible to implement this TODO:
```
  // TODO: if okThrow, permit arguments to the newClass to be non-deterministic (don't add
  // those to purityResult), but still don't permit them to have side effects.
```

If you don't want to do that right now, that's fine.  I was just confused about why there was a long comment about throw expressions in `visitNewClass`.

mernst commented:
@smillst, I added some comments indicating the desirability of moving the code to a different method in the future.  More importantly, I added a check of whether the constructor itself is `@SideEffectFree`.  Could you take another look?  Thanks.

smillst commented on dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java:
```diff
@@ -183,15 +183,55 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void ignore) {
+                purityResult.addNotSEFreeReason(node, "object.creation");
             }
+
+            // TODO: if okThrow, permit arguments to the newClass to be non-deterministic (don't add
```
`okThrow` -> `okThrowDeterministic`
